hub/core/compression/base.py

```diff
@@ -0,0 +1,139 @@
+from typing import Union
+from abc import ABC, abstractmethod  # type: ignore
+import numcodecs  # type: ignore
+from numcodecs.abc import Codec  # type: ignore
+
+import numpy as np
+
+
+class BaseNumCodec(ABC):
+    """Base class for numcodec compressors"""
+
+    @abstractmethod
+    def encode(self, input: np.ndarray) -> bytes:
+        pass
+
+    @abstractmethod
+    def decode(self, bytes: bytes) -> Union[np.ndarray, bytes]:
+        pass
+
+    @property
+    @abstractmethod
+    def __name__(self):
+        pass
+
+
+class BaseImgCodec(ABC, Codec):
+    """Base class for image codecs"""
+
+    def __init__(self, single_channel: bool = True) -> None:
+        self.single_channel = single_channel
+        self._msgpack = numcodecs.MsgPack()
```

Comment:
do we need a `MsgPack` instance for every single compressor object? can we have 1 for the entire hub 2.0 codebase?

Reply:
im asking because i didn't know this existed and it would be cool to use for the index map

Reply:
I don't think that this will work for sure because if there will be > 1 MsgPack encoded objects at the same time (especially if using multiprocessing), it will not be clear which one should be decoded.

Reply:
Not sure where this should go. To `constants.py`?